src/solver/optimisation/opt_gestion_second_membre_cas_lineaire.cpp

```diff
@@ -31,206 +31,435 @@
-                                                     int PremierPdtDeLIntervalle,
-                                                     int DernierPdtDeLIntervalle,
-                                                     int NumeroDeLIntervalle,
-                                                     const int optimizationNumber)
+static void shortTermStorageBalance(const ::ShortTermStorage::AREA_INPUT& shortTermStorageInput,
+                                    ConstraintBuilder& constraintBuilder)
 {
-    int weekFirstHour = problemeHebdo->weekInTheYear * 168;
-
-    PROBLEME_ANTARES_A_RESOUDRE* ProblemeAResoudre = problemeHebdo->ProblemeAResoudre.get();
-
-    std::vector<double>& SecondMembre = ProblemeAResoudre->SecondMembre;
-
-    std::vector<double*>& AdresseOuPlacerLaValeurDesCoutsMarginaux
-      = ProblemeAResoudre->AdresseOuPlacerLaValeurDesCoutsMarginaux;
-
-    int NombreDePasDeTempsDUneJournee = problemeHebdo->NombreDePasDeTempsDUneJournee;
-
-    const std::vector<int>& NumeroDeJourDuPasDeTemps = problemeHebdo->NumeroDeJourDuPasDeTemps;
-    const std::vector<int>& NumeroDeContrainteEnergieHydraulique
-      = problemeHebdo->NumeroDeContrainteEnergieHydraulique;
-    const std::vector<int>& NumeroDeContrainteMinEnergieHydraulique
-      = problemeHebdo->NumeroDeContrainteMinEnergieHydraulique;
-    const std::vector<int>& NumeroDeContrainteMaxEnergieHydraulique
-      = problemeHebdo->NumeroDeContrainteMaxEnergieHydraulique;
-    const std::vector<int>& NumeroDeContrainteMaxPompage
-      = problemeHebdo->NumeroDeContrainteMaxPompage;
-
-    const std::vector<bool>& DefaillanceNegativeUtiliserConsoAbattue
-      = problemeHebdo->DefaillanceNegativeUtiliserConsoAbattue;
-    const std::vector<bool>& DefaillanceNegativeUtiliserPMinThermique
-      = problemeHebdo->DefaillanceNegativeUtiliserPMinThermique;
+    for (const auto& storage : shortTermStorageInput)
+    {
+        const int clusterGlobalIndex = storage.clusterGlobalIndex;
+        constraintBuilder.updateIndex(clusterGlobalIndex);
+        constraintBuilder
+          .include(Variable::ShortTermStorageInjection, 1.0)
+          .include(Variable::ShortTermStorageWithdrawal, -1.0);
+    }
+}
 
-    for (int i = 0; i < ProblemeAResoudre->NombreDeContraintes; i++)
+struct AreaBalance : public IConstraint
+{
+  using IConstraint::IConstraint;
+              
+  void add(int pdt, int pdtHebdo, int pays, int optimizationNumber) override
+  {
+    // TODO improve this
     {
-        AdresseOuPlacerLaValeurDesCoutsMarginaux[i] = nullptr;
+      ConstraintNamer namer(problemeHebdo->ProblemeAResoudre->NomDesContraintes,
+                          problemeHebdo->NamedProblems);
+      namer.UpdateTimeStep(problemeHebdo->weekInTheYear * 168 + pdt);
+      namer.UpdateArea(problemeHebdo->NomsDesPays[pays]);
+      namer.AreaBalance(problemeHebdo->ProblemeAResoudre->NombreDeContraintes);
+    }
+    builder.updateHourWithinWeek(pdt);
```

Comment:
I think we should have a new `builder` for each constraint, it would be more robust.

Actually the build could throw if we try to re-use it, to make 100% sure our constraints creation do not overlap.